### ISSUE TITLE
Root IngressRoutes that define the same FQDN can result in conflicting ingress configuration

### ISSUE THREAD

**Issue by alexbrand** [heptio/contour#601]
Currently, it is possible to define multiple root IngressRoutes that specify the same FQDN. Conflict can arise when two (or more) IngressRoutes define the same FQDN + path combination but point the route to a different service. For example:

```
apiVersion: contour.heptio.com/v1beta1
kind: IngressRoute
metadata: 
  name: numberone
spec: 
  virtualhost:
    fqdn: example.com
  routes: 
    - match: /
      services: 
        - name: service1
          port: 80
---
apiVersion: contour.heptio.com/v1beta1
kind: IngressRoute
metadata: 
  name: numbertwo
spec: 
  virtualhost:
    fqdn: example.com
  routes: 
    - match: /
      services: 
        - name: service2
          port: 80
```

One potential solution is to enforce that a given FQDN can only be used in a single Root IngressRoute. 

xref: https://github.com/heptio/gimbal/issues/209

**Comment by davecheney** [comment]
Yup, I think it was specified in one of the versions of the design doc that
if an fqdn appeared more than once we'd reject _both_ ingressroute
documents with the appropriate status; say, "conflict; fqdn: %q conflicts
with definition in %namespace/%name"

**Comment by alexbrand** [comment]
SGTM. I see two ways of doing this in the dag package:

a) Go over all IngressRoute objects a first time before actually processing them. In this first pass, filter out the ones that conflict because of FQDN reuse. In the second pass, process the IngressRoutes that do not conflict and build the DAG as we do today.

b) Maintain our current approach of iterating over the IngressRoutes once and change our status slice to be a `map[meta]Status`. This is required so that we can update the status of a previously-seen IngressRoute (which was originally deemed valid) in the case we run into a conflicting FQDN. We will also need to remove the VHost from the `_vhosts` and `_svhosts` maps.

The latter option seems messier in my mind. Wondering what your thoughts are on this, and if you can think of any other approaches here. 

Side note: At a higher level, I am starting to think that we might want to split out validation of IngressRoute objects from building the DAG. The only issue here is that there are validation checks that can only be performed while actually building the DAG (e.g. cyclic delegation paths)

**Comment by stevesloka** [comment]
We should also add to this looking for the case of `*`. Only a single one would be allowed across the cluster.

**Comment by alexbrand** [comment]
@stevesloka Do we support wildcards in IngressRoutes?

**Comment by stevesloka** [comment]
We talked about it early on but never came to a conclusion. I know it can make things really complex depending on the routes you set up, but nothing today stops someone from setting `*` or `*.foo.com` for the fqdn I don't think. 

We should either support wildcards (which I think we should) or filter out any fqdn or alias that has `*` in the field by setting the status of that route to error.

**Comment by alexbrand** [comment]
> We talked about it early on but never came to a conclusion.

Got it. That's what I thought too. I couldn't find any reference to it in the design doc, nor any open issue around it.

I just tried creating an IngressRoute with FQDN = `*` and our CRD validation caught it as invalid, so I think we are on the "wildcards are not supported" side of things.

**Comment by stevesloka** [comment]
Ahh ok forgot about that validation. We can leave this issue alone, open a new one and ping @rosskukulinski for thoughts on supporting wildcards since doing that is expanding the original intent of this issue.

**Comment by rosskukulinski** [comment]
I think we'll want to support wildcards eventually, but I think we should take some time and think through all the scenarios (e.g. TLS, etc).  I think there are higher-priority items that we should tackle for 0.6.  I propose opening a new issue to track wildcard domain support, and schedule for 0.7+.

**Comment by alexbrand** [comment]
I have an initial implementation of this. I followed the pre-processing idea I had suggested above. In more details:
1. Before processing the IngressRoute resources in internal/dag/builder.go, build a map that maps FQDNs to the IngressRoutes that are using that fqdn.
2. Iterate over the map. If the IngressRoute slice has more than one IR, build a user-friendly message and mark them as invalid by appending to the Status slice. In the case the the slice has a single IR, append it to another slice for further processing.
